File: SddIA/scripts/qa/fix_tool_process_core.py
# -*- coding: utf-8 -*-
"""fix-tool-process — sandbox Self-Healing (Fase 4)."""

from __future__ import annotations

import json
import os
from pathlib import Path
from typing import Any

from eda_bus_utils import _write_json_atomic, load_radamanto_config
from radamanto_batch_core import set_structure_valid


def sandbox_root(repo: Path) -> Path:
    cfg = load_radamanto_config(repo)
    return repo / (cfg.get("sandbox_root") or ".SddIA/sandbox/")


def materialize_sandbox(repo: Path, entity_id: str, recovery_attempt: int) -> Path:
    root = sandbox_root(repo) / entity_id.replace(":", "_") / str(recovery_attempt)
    root.mkdir(parents=True, exist_ok=True)
    (root / "fix-artifact.md").write_text(
        f"# Fix sandbox\n\nentity={entity_id}\nattempt={recovery_attempt}\n",
        encoding="utf-8",
    )
    return root


def _production_roots(repo: Path) -> list[Path]:
    return [
        (repo / "SddIA" / "tools").resolve(),
        (repo / "SddIA" / "skills").resolve(),
    ]


def assert_sandbox_write(repo: Path, target: Path, sandbox: Path) -> tuple[bool, str | None]:
    if os.environ.get("SDDIA_SANDBOX_STRICT", "1").strip().lower() in ("0", "false", "no"):
        return True, None
    try:
        target.resolve().relative_to(sandbox.resolve())
        return True, None
    except ValueError:
        pass
    for prod in _production_roots(repo):
        try:
            target.resolve().relative_to(prod)
            return False, f"write prohibido en produccion: {target}"
        except ValueError:
            continue
    return True, None


def run_argos_structure_gate(repo: Path, sandbox: Path, entity_id: str) -> dict[str, Any]:
    artifact = sandbox / "fix-artifact.md"
    structure_valid = artifact.is_file() and artifact.stat().st_size > 0
    gate = {
        "structure_valid": structure_valid,
        "entity_id": entity_id,
        "sandbox": str(sandbox.relative_to(repo)).replace("\\", "/"),
        "emitter": "argos",
        "emits_status_restored": False,
    }
    _write_json_atomic(sandbox / "argos_gate.json", gate)
    if structure_valid:
        set_structure_valid(repo, entity_id, True)
    return gate


def process_fix_tool(repo: Path, rel_path: str) -> dict[str, Any]:
    event_path = (repo / rel_path.strip()).resolve()
    if not event_path.is_file():
        return {"ok": False, "error": f"no existe: {rel_path}"}
    try:
        event = json.loads(event_path.read_text(encoding="utf-8"))
    except (OSError, json.JSONDecodeError) as exc:
        return {"ok": False, "error": str(exc)}

    if event.get("event_type") != "Tool_Degraded":
        return {"ok": False, "error": "solo Tool_Degraded inicia fix-tool-process"}

    payload = event.get("payload") or {}
    if not isinstance(payload, dict):
        return {"ok": False, "error": "payload invalido"}
    entity_id = str(payload.get("target_entity_id") or "unknown")
    attempt = int(payload.get("recovery_attempt") or 1)

    sandbox = materialize_sandbox(repo, entity_id, attempt)
    ok_write, err = assert_sandbox_write(repo, sandbox / "fix-artifact.md", sandbox)
    if not ok_write:
        return {"ok": False, "error": err}

    gate = run_argos_structure_gate(repo, sandbox, entity_id)
    return {
        "ok": True,
        "entity_id": entity_id,
        "sandbox": str(sandbox.relative_to(repo)).replace("\\", "/"),
        "argos_gate": gate,
        "status_restored_emitted": False,
    }

File: SddIA/engine/execute-process/src/engine/fix_tool_process_core.rs
//! fix-tool-process — sandbox Self-Healing (paridad `fix_tool_process_core.py`).

use super::fractal_bus::load_radamanto_config;
use super::radamanto_batch_core::set_structure_valid;
use serde_json::{json, Value};
use std::fs;
use std::path::{Path, PathBuf};

fn write_json_atomic(path: &Path, data: &Value) -> Result<(), String> {
    super::eda_bus_topology::write_json_atomic(path, data)
}

fn sandbox_root(repo: &Path) -> Result<PathBuf, String> {
    let cfg = load_radamanto_config(repo)?;
    let rel = cfg
        .get("sandbox_root")
        .and_then(|v| v.as_str())
        .unwrap_or(".SddIA/sandbox/");
    Ok(repo.join(rel.trim_start_matches("./")))
}

fn materialize_sandbox(repo: &Path, entity_id: &str, recovery_attempt: i64) -> Result<PathBuf, String> {
    let root = sandbox_root(repo)?
        .join(entity_id.replace(':', "_"))
        .join(recovery_attempt.to_string());
    fs::create_dir_all(&root).map_err(|e| e.to_string())?;
    fs::write(
        root.join("fix-artifact.md"),
        format!("# Fix sandbox\n\nentity={entity_id}\nattempt={recovery_attempt}\n"),
    )
    .map_err(|e| e.to_string())?;
    Ok(root)
}

fn production_roots(repo: &Path) -> Vec<PathBuf> {
    vec![
        repo.join("SddIA/tools").canonicalize().unwrap_or_else(|_| repo.join("SddIA/tools")),
        repo.join("SddIA/skills")
            .canonicalize()
            .unwrap_or_else(|_| repo.join("SddIA/skills")),
    ]
}

fn assert_sandbox_write(repo: &Path, target: &Path, sandbox: &Path) -> Result<(), String> {
    let strict = !matches!(
        std::env::var("SDDIA_SANDBOX_STRICT")
            .unwrap_or_else(|_| "1".into())
            .trim()
            .to_lowercase()
            .as_str(),
        "0" | "false" | "no"
    );
    if !strict {
        return Ok(());
    }
    if target
        .canonicalize()
        .ok()
        .and_then(|t| sandbox.canonicalize().ok().map(|s| t.starts_with(&s)))
        .unwrap_or(false)
    {
        return Ok(());
    }
    for prod in production_roots(repo) {
        if target
            .canonicalize()
            .ok()
            .and_then(|t| prod.canonicalize().ok().map(|p| t.starts_with(&p)))
            .unwrap_or(false)
        {
            return Err(format!("write prohibido en produccion: {}", target.display()));
        }
    }
    Ok(())
}

fn run_argos_structure_gate(repo: &Path, sandbox: &Path, entity_id: &str) -> Result<Value, String> {
    let artifact = sandbox.join("fix-artifact.md");
    let structure_valid = artifact.is_file() && fs::metadata(&artifact).map(|m| m.len() > 0).unwrap_or(false);
    let gate = json!({
        "structure_valid": structure_valid,
        "entity_id": entity_id,
        "sandbox": sandbox.strip_prefix(repo).unwrap_or(sandbox).to_string_lossy().replace('\\', "/"),
        "emitter": "argos",
        "emits_status_restored": false,
    });
    write_json_atomic(&sandbox.join("argos_gate.json"), &gate)?;
    if structure_valid {
        set_structure_valid(repo, entity_id, true)?;
    }
    Ok(gate)
}

pub fn process_fix_tool(repo: &Path, rel_path: &str) -> Value {
    match process_fix_tool_inner(repo, rel_path) {
        Ok(v) => v,
        Err(e) => json!({"ok": false, "error": e}),
    }
}

fn process_fix_tool_inner(repo: &Path, rel_path: &str) -> Result<Value, String> {
    let event_path = repo.join(rel_path.trim());
    if !event_path.is_file() {
        return Err(format!("no existe: {rel_path}"));
    }
    let event: Value =
        serde_json::from_str(&fs::read_to_string(&event_path).map_err(|e| e.to_string())?)
            .map_err(|e| e.to_string())?;

    if event.get("event_type").and_then(|v| v.as_str()) != Some("Domain_Entity_Degraded") {
        return Err("solo Domain_Entity_Degraded inicia fix-tool-process".into());
    }
    let payload = event.get("payload").cloned().unwrap_or(json!({}));
    if !payload.is_object() {
        return Err("payload invalido".into());
    }
    let entity_id = payload
        .get("entity_id")
        .or_else(|| payload.get("target_entity_id"))
        .and_then(|v| v.as_str())
        .unwrap_or("unknown")
        .to_string();
    let entity_type = payload
        .get("entity_type")
        .and_then(|v| v.as_str())
        .unwrap_or("tool");
    if entity_type != "tool" {
        return Ok(json!({
            "ok": true,
            "skipped": true,
            "reason": format!("entity_type={entity_type} fuera de alcance fix-tool-process"),
            "entity_id": entity_id,
        }));
    }
    let attempt = payload
        .get("recovery_attempt")
        .and_then(|v| v.as_i64())
        .unwrap_or(1);
    let sandbox = materialize_sandbox(repo, &entity_id, attempt)?;
    assert_sandbox_write(repo, &sandbox.join("fix-artifact.md"), &sandbox)?;
    let gate = run_argos_structure_gate(repo, &sandbox, &entity_id)?;
    Ok(json!({
        "ok": true,
        "entity_id": entity_id,
        "sandbox": sandbox.strip_prefix(repo).unwrap_or(&sandbox).to_string_lossy().replace('\\', "/"),
        "argos_gate": gate,
        "status_restored_emitted": false,
    }))
}
